subsys/usb/class/audio/audio.c

```diff
@@ -0,0 +1,958 @@
+ */
+static bool is_interface_valid(struct usb_audio_dev_data_t *audio_dev_data,
+			       u8_t interface)
+{
+	const struct cs_ac_interface_descriptor_header *header;
+
+	header = audio_dev_data->header_descr;
+	u8_t desc_iface = 0;
+
+	for (size_t i = 0; i < header->bInCollection; i++) {
+		desc_iface = header->baInterfaceNr[i];
+		if (desc_iface == interface) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * @brief Helper funciton for getting the audio_dev_data by the interface
+ *	  number.
+ *
+ * This function searches through all audio devices the one with given
+ * interface number and returns the audio_dev_data structure for this device.
+ *
+ * @param [in] interface USB Audio interface addressed by the request.
+ *
+ * @return audio_dev_data for given interface, NULL if not found.
+ */
+static struct usb_audio_dev_data_t *get_audio_dev_data_by_interface(
+								u8_t interface)
+{
+	struct usb_dev_data *dev_data;
+	struct usb_audio_dev_data_t *audio_dev_data;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&usb_audio_data_devlist, dev_data, node) {
+		audio_dev_data = CONTAINER_OF(dev_data,
+					struct usb_audio_dev_data_t, common);
+
+		if (is_interface_valid(audio_dev_data, interface)) {
+			return audio_dev_data;
+		}
+	}
+	return NULL;
+}
+
+/**
+ * @brief Handler for feature unit requests.
+ *
+ * This function handles feature unit specific requests.
+ * If request is properly served 0 is returned. Negative errno
+ * is returned in case of an error. This leads to setting stall on IN EP0.
+ *
+ * @param [in]  dev_data USB audio device data.
+ * @param [in]  pSetup   Information about the executed request.
+ * @param [in]  len      Size of the buffer.
+ * @param [out] data     Buffer containing the request result.
+ *
+ * @return 0 if succesfulf, negative errno otherwise.
+ */
+static int handle_feature_unit_req(struct usb_audio_dev_data_t *audio_dev_data,
+				   struct usb_setup_packet *pSetup,
+				   s32_t *len,
+				   u8_t **data)
+{
+	enum usb_audio_feature_unit_control_selectors cs;
+	struct usb_audio_feature_unit_evt evt;
+	struct feature_unit_descriptor *fu;
+	u8_t device;
+	u8_t fu_id;
+	u8_t ch;
+
+	fu_id = ((pSetup->wIndex) >> 8) & 0xFF;
+	fu = get_feature_unit(audio_dev_data, &device, fu_id);
+	ch = (pSetup->wValue) & 0xFF;
+	cs = ((pSetup->wValue) >> 8) & 0xFF;
+
+	LOG_DBG("CS: %d, CN: %d, len: %d", cs, ch, *len);
+
+	/* Error checking */
+	if (!(BIT(cs) & (get_controls(fu) << 1UL))) {
+		return -EINVAL;
+	} else if (ch >= audio_dev_data->ch_num[device] &&
+		   ch != 0xFF) {
+		return -EINVAL;
+	}
+
+	switch (cs) {
+	case USB_AUDIO_FU_MUTE_CONTROL:
+		switch (pSetup->bRequest) {
+		case USB_AUDIO_SET_CUR:
+			if (ch == 0xFF) {
+				memcpy(&audio_dev_data->controls[device][POS(MUTE, 0, audio_dev_data->ch_num[device])],
+				       *data, LEN(audio_dev_data->ch_num[device], MUTE));
+			} else {
+				audio_dev_data->controls[device][ch] =
+							(*data[0] != 0);
+			}
```

Comment:
Why not use local variables:
```
	u8_t *controls = audio_dev_data->controls[device];
	u8_t ch_num = audio_dev_data->ch_num[device];
```
and shorten the above to:
```
			if (ch == 0xFF) {
				memcpy(&controls[POS(MUTE, 0, ch_num)],
				       *data, LEN(ch_num, MUTE));
			} else {
				controls[POS(MUTE, ch, ch_num)] = (*data)[ch];
			}
```
?
BTW `*data[0]` is wrong in this context. First of all it always gets the value of the first channel instead of the channel `ch`. But to avoid confusion, `*data` should be in parentheses because `*data[i]` here is equivalent to `(*data)[i]` only when `i` is 0.
And don't you need to bother with `*len` here?

Reply:
> BTW *data[0] is wrong in this context. First of all it always gets the value of the first channel instead of the channel ch.

It does not. If `ch != 0xFF` the data for addressed channel are stored starting from index 0 of data pointer. For example if `ch = 2` then data are under `(*data)[0]` . The len should remain '1' (in case of MUTE) that's why I omitted this but worthy to bring it back.

Reply:
I will use locals for sure, will make the code much easier to read.

Reply:
> > First of all it always gets the value of the first channel instead of the channel ch.
> 
> It does not. If ch != 0xFF the data for addressed channel are stored starting from index 0 of data pointer.

Right. My mistake. But why use an index then when the pointed value is not an array?

Reply:
Uhm.. my bad, will change to `**data` as it should be from the beginning.